src/main/java/com/iridium/iridiumskyblock/managers/IslandManager.java

```diff
@@ -283,6 +284,69 @@ private void deleteIslandBlocks(Island island, World world, int y, CompletableFu
         }
     }
 
+    public CompletableFuture<Void> regenerateTerrain(Island island) {
+        return CompletableFuture.runAsync(() -> {
+            List<CompletableFuture<Void>> completableFutures = Arrays.asList(
+                    regenerateTerrain(island, getWorld(World.Environment.NORMAL)),
+                    regenerateTerrain(island, getWorld(World.Environment.NETHER)),
+                    regenerateTerrain(island, getWorld(World.Environment.THE_END))
+            );
+            completableFutures.forEach(CompletableFuture::join);
+        });
+    }
+
+    private CompletableFuture<Void> regenerateTerrain(Island island, World world) {
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        if (world == null) {
+            completableFuture.complete(null);
+        } else {
+            Bukkit.getScheduler().runTask(IridiumSkyblock.getInstance(), () -> regenerateTerrain(island, world, world.getMaxHeight(), completableFuture, 0));
+        }
+        return completableFuture;
+    }
+
+    public void regenerateTerrain (Island island, World world, int y, CompletableFuture<Void> completableFuture, int delay) {
+
+        if (world == null) return;
+        Location pos1 = island.getPosition1(world);
+        Location pos2 = island.getPosition2(world);
+
+        if(!(IridiumSkyblock.getInstance().getDefaultWorldGenerator(world.getName(), null) instanceof VoidGenerator)) {
+
+            String worldName = "plugins/iridiumskyblock" + "/" + "regenWorlds" + "/" + world.getName() + "_regen";
+            if (Bukkit.getWorld(worldName) == null) {
+
+                WorldCreator worldCreator = new WorldCreator(worldName)
+                        .generator(IridiumSkyblock.getInstance().getDefaultWorldGenerator(world.getName(), null))
+                        .environment(world.getEnvironment())
+                        .seed(world.getSeed());
+
+                worldCreator.createWorld();
+            }
+
+            World regenWorld = Bukkit.getWorld(worldName);
+
+            for (int x = pos1.getBlockX(); x <= pos2.getBlockX(); x++) {
+                for (int z = pos1.getBlockZ(); z <= pos2.getBlockZ(); z++) {
+                    Block blockA = regenWorld.getBlockAt(x, y, z);
+                    Block blockB = world.getBlockAt(x, y, z);
+
+                    blockB.setType(blockA.getType(), false);
+                }
+            }
+
+            if (y <= LocationUtils.getMinHeight(world)) {
+                completableFuture.complete(null);
+            } else {
+                if (delay < 1) {
+                    regenerateTerrain(island, world, y - 1, completableFuture, delay);
+                } else {
+                    Bukkit.getScheduler().runTaskLater(IridiumSkyblock.getInstance(), () -> regenerateTerrain(island, world, y - 1, completableFuture, delay), delay);
+                }
+            }
+        }
+    }
```

Comment:
Potentially we can utilise the Paster class, which should already have this code in

Reply:
i can look into this, but i separated it into its own method because it will only run this if it detects we're using a terrain generator